x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authz/RBACEngine.java

```diff
@@ -500,15 +500,21 @@ GetUserPrivilegesResponse buildUserPrivilegesResponseObject(Role userRole) {
         return new GetUserPrivilegesResponse(cluster, conditionalCluster, indices, application, runAs);
     }
 
-    static List<String> resolveAuthorizedIndicesFromRole(Role role, String action, Map<String, IndexAbstraction> aliasAndIndexLookup) {
-        Predicate<String> predicate = role.allowedIndicesMatcher(action);
+    static List<String> resolveAuthorizedIndicesFromRole(Role role, RequestInfo requestInfo, Map<String, IndexAbstraction> lookup) {
+        Predicate<String> predicate = role.allowedIndicesMatcher(requestInfo.getAction());
+
+        // do not include data streams for actions that do not operate on data streams
+        TransportRequest request = requestInfo.getRequest();
+        boolean includeDataStreams = (request instanceof IndicesRequest) && ((IndicesRequest) request).includeDataStreams();
 
         List<String> indicesAndAliases = new ArrayList<>();
         // TODO: can this be done smarter? I think there are usually more indices/aliases in the cluster then indices defined a roles?
-        for (Map.Entry<String, IndexAbstraction> entry : aliasAndIndexLookup.entrySet()) {
+        for (Map.Entry<String, IndexAbstraction> entry : lookup.entrySet()) {
             String aliasOrIndex = entry.getKey();
             if (predicate.test(aliasOrIndex)) {
-                indicesAndAliases.add(aliasOrIndex);
+                if (entry.getValue().getType() != IndexAbstraction.Type.DATA_STREAM || includeDataStreams) {
+                    indicesAndAliases.add(aliasOrIndex);
+                }
             }
```

Comment:
I gave it more thought than it probably warrants.

Overall I am OK with the change, but I would suggest we move this whole logic( `if (entry.getValue().getType() != IndexAbstraction.Type.DATA_STREAM || includeDataStreams) {`) to https://github.com/elastic/elasticsearch/blob/2925056e78d87069aa4dcdf1c7fb398830906315/x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authz/IndicesAndAliasesResolver.java#L113

This is the place where we deal with the entities (open/close indices/aliases) behind the "authorised" names, and I believe it would help us keeping complexity under control in an already messy spaghetti.

Reply:
Would moving the logic to `IndicesAndAliasesResolver` still achieve the desired result? The problem is that all requests that implement `IndicesRequest.Replaceable` have their requested indices replaced with the result of the `loadAuthorizedIndices` call and this causes problems for the requests such as `GET /_alias` that do not (and should not) operate on data streams.

Reply:
What I mean is to move the whole processing on `loadAuthorizedIndices`, as you've proposed, over to the method in `IndicesAndAliasesResolver`. I think the outcome is the same, but it feels to me like a better place for these kind of processing on index names based on requests. Makes sense?

Reply:
Ah, thanks for the clarification. I will make that change.